### DISCUSSION TITLE
Hiding connection strings and secrets

### DISCUSSION THREAD

**Discussion by cilerler** [dotnet/interactive#3789] [Q&A] [answered]
While working with a polyglot notebook, it’s necessary to avoid displaying the SQL database connection string. 
Directly storing it in a variable reveals it in the variable list. How can this string be safely concealed? 

Additionally, can the same approach be used to protect other sensitive secrets as well?

```notebook
#!meta

{"kernelInfo":{"defaultKernelName":"csharp","items":[{"aliases":[],"languageName":"csharp","name":"csharp"},{"aliases":[],"languageName":"T-SQL","name":"sql-Integration_MyCompany"}]}}

#!csharp

#r "nuget:Microsoft.DotNet.Interactive.SqlServer,*-*"
var connectionString = System.Environment.GetEnvironmentVariable("integration_smartpocure_sql");

#!csharp

#!connect mssql --kernel-name Integration_MyCompany --connection-string @connectionString

#!markdown

Test SQL Connection with a Query

#!sql-Integration_MyCompany

SELECT @@VERSION;

```

**Comment by jonsequitur** [comment, marked as answer]
You can securely store secret values via magic commands, as described here: 

https://github.com/dotnet/interactive/blob/main/docs/input-prompts.md#saving-a-value

Please let us know if this works for you and if you run into any issues or usability problems.

**Reply by cilerler** [reply]
In case it helps someone else, this is the workaround I ended up using.

> I'm looking into the bug, which I was able to reproduce once, but can't reproduce again after a VS Code restart.
> 
> Side note: You don't need to set a variable to use the secret management feature. You can collapse the two magic command steps down to this:
> 
> ```
> #!connect mssql --kernel-name Integration_MyCompany --connection-string @password:{"saveAs":"mycompany.integration.sql", "prompt":"Please provide a connection string."}
> ```

**Comment by cilerler** [comment]
Thank you!  
I was able to save it using the following command:

```shell
#!set --name myCompanyIntegrationSql --value @password:{"saveAs":"myCompany.integration.sql", "prompt":"Please provide a connection string."}
```

While this resolves the issue I was facing, the experience could be significantly improved with a feature like `polyglot-notebook.kernelSecretVariables`, similar to `polyglot-notebook.kernelEnvironmentVariables` in the settings. This would make it easier to configure secrets without worrying about accidental exposure, while keeping the use of PowerShell Secret Store optional.

**Reply by jonsequitur** [reply]
An important goal of this design was to discourage storing secrets in plain text. The VS Code config has the additional danger in this case of being copied to other machines if you've enabled settings sync, so storing secrets there would be risky.

**Reply by cilerler** [reply]
Got it—thanks for clarifying. I really appreciate your help. Have a great year, and thanks again!

**Reply by jonsequitur** [reply]
You're welcome!

**Reply by cilerler** [reply]
I think I spoke too soon. While the method did allow me to securely store the variable, I'm unable to pass it to the `connect` command. Below is the code snippet and the error I’m receiving:

## Code Snippet
```csharp
#!set --name myCompanyIntegrationSql --value @password:{"saveAs":"myCompany.integration.sql", "prompt":"Please provide a connection string."}
```

```csharp
#!connect mssql --kernel-name Integration_MyCompany --connection-string @myCompanyIntegrationSql
```

## Error
```none
Error: System.Exception: Format of the initialization string does not conform to specification starting at index 0.
   at Microsoft.DotNet.Interactive.SqlServer.MsSqlKernel.ConnectAsync()
   at Microsoft.DotNet.Interactive.SqlServer.MsSqlKernelConnector.CreateKernelAsync(String kernelName)
   at Microsoft.DotNet.Interactive.SqlServer.ConnectMsSqlDirective.ConnectKernelsAsync(ConnectMsSqlKernel connectCommand, KernelInvocationContext context)
   at Microsoft.DotNet.Interactive.CompositeKernel.ConnectKernel[TCommand](TCommand command, ConnectKernelDirective`1 connectDirective, KernelInvocationContext context) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive\CompositeKernel.cs:line 253
   ...
   ...
   at Microsoft.DotNet.Interactive.KernelCommandPipeline.SendAsync(KernelCommand command, KernelInvocationContext context) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive\KernelCommandPipeline.cs:line 41
```

## Additional

If I convert the value with:

```csharp
var connectionString = myCompanyIntegrationSql.GetClearTextPassword();
```

and then call:

```csharp
#!connect mssql --kernel-name Integration_MyCompany --connection-string @connectionString
```

it does work, but that makes the connection string appear in clear text.

Any suggestion?

**Reply by cilerler** [reply]
Closing this discussion since it partially addresses the question. 
The remaining part is now documented as a bug/feature request at https://github.com/dotnet/interactive/issues/3793.

Thanks again.  🤗

**Reply by jonsequitur** [reply]
I'm looking into the bug, which I was able to reproduce once, but can't reproduce again after a VS Code restart. 

Side note: You don't need to set a variable to use the secret management feature. You can collapse the two magic command steps down to this:

```
#!connect mssql --kernel-name Integration_mycompany --connection-string @password:{"saveAs":"mycompany.integration.sql", "prompt":"Please provide a connection string."}
```

**Reply by cilerler** [reply]
Thank you again! It worked like a charm—it accepted it, and I no longer see the value in the variables list. 
That said, I believe the bug should remain open since `connect mssql` still doesn’t accept `passwordstring`. Do you agree?

**Reply by jonsequitur** [reply]
Yeah, the issue you opened is definitely a bug.

**Reply by cilerler** [reply]
Thanks. Have a healthy, and happy new year. 🤗
